Added multi cartpole example

veds12 opened the pull request:
Description:

* The multi-cartpole grid example consists of 9 cartpole robots each present in one block of a 3x3 grid
* Each cartpole robot functions independently of each other
* The supervisor communicated with each of these 9 robots through separate emitter-receiver channels, i.e. there are 9 pairs of emitter-receiver channels
* The robots are trained using Proximal Policy Optimisation
* The supervisor controller can work for more / less number of robots by changing the value of the parameter `num_robots` while initializing it. The robot controller is agnostic to the number of robots. Hence, this world can be easily extended to different number of robots.

TO DO:

- [x] 9 cartpole grid env
- [x] Simultaneous communication using emmiter-receiver scheme working
- [x] Training
- [x] Clean the code 
- [x] Change communication scheme 
- [x] Move all the SolidBox nodes to a group node
- [x] Add reward plotting

veds12 committed:
added multi-cartpole grid env

veds12 committed:
training successful

veds12 committed:
cleaned controllers

veds12 committed:
moved multi-cartpole to examples

veds12 committed:
deepbots update

veds12 committed:
update comm scheme

veds12 committed:
updateed multi-cartpole

veds12 committed:
dummy commit

veds12 committed:
fixed multicartpole

veds12 committed:
Merge pull request #1 from veds12/dev

added multi-cartpole grid env

veds12 committed:
Merge branch 'dev' of https://github.com/aidudezzz/deepworlds into dev
fetch updates

veds12 committed:
Merge branch 'dev' of https://github.com/eellak/gsoc2021-deepbots-evo into dev
fetch updates from upstream

KelvinYang0320 commented:
@veds12 Thank you for this great and interesting example.:thumbsup: 
I will do a code review as soon as possible.

Could you move all SolidBox nodes to [a group node](https://cyberbotics.com/doc/reference/group).
By doing so, I think it will be easier for end-users to understand the [scene tree](https://cyberbotics.com/doc/guide/the-scene-tree) in this example.

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/PPO_runner.py:
```diff
@@ -0,0 +1,88 @@
+    agent = PPOAgent(supervisorEnv.observationSpace, supervisorEnv.actionSpace)
+
+    episodeCount = 0
+    episodeLimit = 2000
```
This constant can be set on top of supervisor_manager/supervisor_manager.py (something like `EPISODE_LIMIT = 2000`)  in order to be visible to the end user.
See [panda example](https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_manager.py#L7-L11) as suggested by @ManosMagnus during [pull request](https://github.com/aidudezzz/deepworlds/pull/33#discussion_r603875958).
https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_manager.py#L7-L11

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/PPO_runner.py:
```diff
@@ -0,0 +1,88 @@
+            # done condition
+            newState, reward, done, info = supervisorEnv.step([*selectedActions])
+
+            # Save the current state transitions from both robots in agent's memory
```
I'm not sure if I get what you mean.
"Both" is used for emphasis to refer to two robots, but there are 9 robots in this example.

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/PPO_runner.py:
```diff
@@ -0,0 +1,88 @@
+            state = newState  # state for next step is current step's newState
+
+        if supervisorEnv.test:  # If test flag is externally set to True, agent is deployed
+            break
```
From my understanding, the code in supervisor_manager/supervisor_controller.py doesn't change it to `self.ture = True`.
We can remove these lines?

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/supervisor_controller.py:
```diff
@@ -0,0 +1,286 @@
+
+        self.messageReceived = None                     # Variable to save the messages received from the robots
+
+        self.stepsPerEpisode = 200                      # Number of steps per episode
```
This constant can be set on top of supervisor_manager/supervisor_manager.py (something like `STEPS_PER_EPISODE = 200`)  in order to be visible to the end user.
See [panda example](https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_manager.py#L7-L11) as suggested by @ManosMagnus during [pull request](https://github.com/aidudezzz/deepworlds/pull/33#discussion_r603875958).
https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_manager.py#L7-L11

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/supervisor_manager.py:
```diff
@@ -0,0 +1,6 @@
+More runners for discrete RL algorithms can be added here.
+"""
+import PPO_runner
+
```
Add constants here.
See [panda example](https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_manager.py#L7-L11) as suggested by @ManosMagnus during [pull request](https://github.com/aidudezzz/deepworlds/pull/33#discussion_r603875958).
https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/panda/panda_goal_reaching/controllers/robot_supervisor_manager/robot_supervisor_manager.py#L7-L11

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/PPO_runner.py:
```diff
@@ -0,0 +1,88 @@
+        if not solved:
+            print("Task is not solved, deploying agent for testing...")
+        elif solved:
+            print("Task is solved, deploying agent for testing...")
```
Can you use `plotData` imported from `utilities` to plot a [Reward per episode plot](https://github.com/aidudezzz/deepworlds/tree/dev/examples/cartpole) after the task is solved.
See plotData in carpole:
https://github.com/aidudezzz/deepworlds/blob/95af837c896a859c95835e7b41e8b79acba94aae/examples/cartpole/cartpole_discrete/controllers/robot_supervisor_manager/PPO_runner.py#L63-L64

KelvinYang0320 reviewed (changes requested):
Thank you for the great code quality! 
I've left a few comments.
Also, please add a readme with showcase & plot to this example as we have done for each example in deepworlds.
Cheers!

veds12 commented:
Thanks for the review @KelvinYang0320. I'll make the changes and update the PR as soon as possible

veds12 committed:
added training gif

veds12 committed:
updated multi-cartpole

veds12 committed:
Merge branch 'dev' of https://github.com/aidudezzz/deepworlds into dev
updates from deepworlds-dev

veds12 committed:
added a group node for boundary boxes

veds12 committed:
added plotting at the end of training

veds12 commented:
@KelvinYang0320 I have addressed the requested changes. Please take a look

KelvinYang0320 commented:
@veds12 Did you use wandb in multi_cartpole?
I think you added `evolutionary/cartpole_evolutionary/controllers/supervisor_manager/wandb` by mistake.

> Also, please add a readme with showcase & plot to this example as we have done for each example in deepworlds.

Please add a readme [here](https://github.com/eellak/gsoc2021-deepbots-evo/tree/dev/examples/multi_cartpole).

Could you also fix the coordinate system (NUE → ENU) in this PR? Thank you!
* https://github.com/aidudezzz/deepworlds/issues/58
* [Webots R2022a: Conversion to FLU/ENU](https://www.cyberbotics.com/doc/blog/Webots-2022-a-release?tab-language=python#conversion-to-fluenu)
* ![擷取](https://user-images.githubusercontent.com/49781698/175850151-4ecfcbac-db07-4a37-bc2d-866d85e61c7f.PNG)

veds12 commented:
@KelvinYang0320 Yeah I left the wandb logs by mistake. Forgot to update the gitignore. I'll make the other changes.
Thanks!

veds12 committed:
added readme & removed trash

veds12 committed:
updated coordinate system in the controller

veds12 commented:
- I have added a readme and changed the coordinate system (in the controller code).
- I tried training the example but for some reason it's taking too long to converge / not converging at all. Currently looking into it. I'll add the training plot when I fix this.

veds12 committed:
Merge branch 'dev' of https://github.com/aidudezzz/deepworlds into dev
fetch updates

KelvinYang0320 commented on examples/multi_cartpole/controllers/supervisor_manager/supervisor_controller.py:
```diff
@@ -0,0 +1,283 @@
+                poleAngle = [0.0 for _ in range(self.num_robots)]
+
+        # Angular velocity x of endpoint
+        endpointVelocity = [normalizeToRange(self.poleEndpoint[i].getVelocity()[3], -1.5, 1.5, -1.0, 1.0, clip=True) for i in range(self.num_robots)]
```
This might be the reason why it's taking too long to converge / not converging.
* https://github.com/aidudezzz/deepworlds/pull/68
```suggestion
        endpointVelocity = [normalizeToRange(self.poleEndpoint[i].getVelocity()[4], -1.5, 1.5, -1.0, 1.0, clip=True) for i in range(self.num_robots)]
```

KelvinYang0320 reviewed (changes requested):
@veds12 I have left a comment about the training issue.
Could you format your code with PEP8?
I will do a review again after that.
Thank you!

tsampazk commented:
> @veds12 I have left a comment about the training issue. Could you format your code with PEP8? I will do a review again after that. Thank you!

Hey @veds12, just a clarification because we have been discussing it with @KelvinYang0320, i think that the only differences from PEP8 are that we use a line limit of 120, and prefer to use mixedCase (sometimes also called lowerCamelCase) instead of snake_case in variable names.

veds12 commented:
@tsampazk thanks for the clarification! I'll address the changes this weekend
